daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala

```diff
@@ -521,9 +526,18 @@ object ScenarioLedger {
       processNodes(Right(ledgerData), List(None -> richTr.transaction.roots.toList))
 
     mbCacheAfterProcess.map { cacheAfterProcess =>
-      richTr.disclosures(cacheAfterProcess.coidToNodeId).foldLeft(cacheAfterProcess) {
-        case (cacheP, (nodeId, witnesses)) =>
-          cacheP.updateLedgerNodeInfo(nodeId)(_.addObservers(witnesses.map(_ -> trId).toMap))
+      // NOTE(MH): Since `addDisclosures` is biased towards existing
+      // disclosures, we need to add the "stronger" explicit ones first.
+      val cacheWithExplicitDisclosures =
+        richTr.explicitDisclosure.foldLeft(cacheAfterProcess) {
+          case (cacheP, (nodeId, witnesses)) =>
+            cacheP.updateLedgerNodeInfo(EventId(richTr.transactionId, nodeId))(
+              _.addDisclosures(witnesses.map(_ -> Disclosure(since = trId, explicit = true)).toMap))
+        }
+      richTr.globalImplicitDisclosure.foldLeft(cacheWithExplicitDisclosures) {
+        case (cacheP, (coid, witnesses)) =>
```

Comment:
Shouldn’t `witnesses` be called `divulgees` here?

Reply:
That's a good idea. Thanks. I blindly copied the old code...